### PR TITLE
Relax requirements about module names to also allow dash characters

### LINKED ISSUE #162: [FR] Relax validation of `py-modules`
> The pip-run project defines a [py-module that has a dash in the name](https://github.com/jaraco/pip-run/blob/ee5cbbd03d19e7b33effb7d286e79cfcc45c53d8/setup.cfg#L18). It does so intentionally in order to make it possible to type `python -m pip-run` without having to enter any chorded characters (`-` is roughly 1/3 as expensive as `_`).
> 
> After converting the config to pyproject.toml, however, this value is rejected by policy.
> 
> ```
> ValueError: invalid pyproject.toml config: `tool.setuptools.py-modules[0]`.
> configuration error: `tool.setuptools.py-modules[0]` must be python-module-name
> Backend: run command get_requires_for_build_editable with args {'config_settings': None}
> configuration error: `tool.setuptools.py-modules[0]` must be python-module-name
> GIVEN VALUE:
>     "pip-run"
> 
> OFFENDING RULE: 'format'
> 
> DEFINITION:
>     {
>         "type": "string",
>         "format": "python-module-name"
>     }
> ```
> 
> I can understand why "python-module-name" seemed like a good reasonableness check, but it turns out to be backward incompatible and unnecessarily constraining against a use-case like above.
> 
> Can we back out this change and make it lenient as it was with setup.cfg?

@jaraco in https://github.com/pypa/setuptools/issues/4316

### PR DESCRIPTION
**Pull request by abravalheri**
Closes #162.

See details of the discussion in https://github.com/pypa/setuptools/issues/4316#issuecomment-2068123376.

### CODE REVIEW

File: src/validate_pyproject/formats.py [hunk 1]
```diff
@@ -289,6 +289,25 @@ def python_module_name(value: str) -> bool:
...
+    The following are disallowed:
+
+    * names starting/ending in dashes,
+    * names ending in ``-stubs`` (potentially collide with :obj:`pep561_stub_name`).
+    """
+    if value.startswith("-") or value.endswith("-"):
+        return False
+    if value.endswith("-stubs"):
+        return False  # Avoid collision with PEP 561
+    return python_module_name(value.replace("-", "_"))
```
Comment (abravalheri) [inline src/validate_pyproject/formats.py:308]: @jaraco would this be enough to support the use case?

I am trying to remain still a bit conservative...

**Comment by abravalheri** [comment]
@henryiii, I don't know what are your thoughts on this.

If you have any (conceptual) concerns, maybe the best place to discuss would be https://github.com/pypa/setuptools/issues/4316?

File: src/validate_pyproject/formats.py (hunk 1 above)
Comment (abravalheri) [inline src/validate_pyproject/formats.py:308]: @jaraco would this be enough to support the use case?

I am trying to remain still a bit conservative...

Comment (jaraco) [reply src/validate_pyproject/formats.py:308]: Yes, this seems perfectly reasonable to me. Thanks!

**Pull request merged**